x-pack/elastic-agent/pkg/capabilities/capabilities.go

```diff
@@ -0,0 +1,97 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package capabilities
+
+import (
+	"errors"
+	"os"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/logger"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/status"
+)
+
+// Capability provides a way of applying predefined filter to object.
+// It's up to capability to determine if capability is applicable on object.
+type Capability interface {
+	// Apply applies capabilities on input and returns true if input should be completely blocked
+	// otherwise, false and updated input is returned
+	Apply(interface{}) (interface{}, error)
```

Comment:
I still would say the interface is doing to much, while not really abstracting things.

The input and output types are `interface{}`, that is the internal representation of the objects are not really abstracted away behind coherent interfaces. Instead we rely on internal knowledge of the potential structures to be passed from different call sides. Due to the need to downcast and analyze the internal structure, we fail to have proper abstractions here. Instead we allow for potential bugs being introduced in the future in case we change the internal representation for e.g. Inputs to a proper go-type without finding and updating the right capabilities.


The `Apply` method indicates that we are doing a transformation. Yet we only want to filter.  Would it make sense to change the interface to act more like a predicate and have the code to move through the configured structures in some other central place?
My understanding is that we have the 'walking' on the structure already in the transpiler, why should we replicate the walking in the capabilities?

E.g.

```
type Capability interface {
  // TestFeature returns nil if the kind and typeName are accepted by the capability.
  // ErrDeny is returned if the capability rejects the current feature.
  TestFeature(kind ComponentKind, typeName string) (err error)
}

var ErrDeny = errors.New("deny")

type ComponentKind int

const (
  ComponentInput ComponentKind = iota
  ComponentOutput
)
```


Alternatively (in case we want transformations), we might want to pass the root configuration object to `Apply`. In that case only the implementation is abstracted away, but the input type is known. If we change the go type for inputs/outputs to proper structures, the actual implementation of the input capability should fail to compile.